File: php/src/PokemonGoCalc/Models/Iv.php
<?php

declare(strict_types=1);

namespace PokemonGoCalc\Models;

use InvalidArgumentException;
use PokemonGoCalc\Constants\IvConstant;

/**
 * 個体値（Individual Values）.
 */
final readonly class Iv
{
    public function __construct(
        public int $attack,
        public int $defense,
        public int $stamina,
    ) {
        self::validate($attack, 'attack');
        self::validate($defense, 'defense');
        self::validate($stamina, 'stamina');
    }

    /**
     * 合計値を返す.
     */
    public function total(): int
    {
        return $this->attack + $this->defense + $this->stamina;
    }

    /**
     * パーセンテージを返す（0-100%）.
     */
    public function percentage(): float
    {
        return ($this->total() / IvConstant::MAX_TOTAL)
            * IvConstant::PERCENTAGE_MULTIPLIER;
    }

    public function __toString(): string
    {
        return "{$this->attack}/{$this->defense}/{$this->stamina}";
    }

    private static function validate(int $value, string $name): void
    {
        if ($value < IvConstant::IV_MIN || $value > IvConstant::IV_MAX) {
            throw new InvalidArgumentException(
                sprintf(
                    '%sは%d〜%dの範囲で指定してください（入力値: %d）',
                    $name,
                    IvConstant::IV_MIN,
                    IvConstant::IV_MAX,
                    $value,
                ),
            );
        }
    }
}

File: api/src/domain/value_objects/iv.py
"""個体値 (Individual Values) Value Object."""

from pydantic import BaseModel, Field

IV_MIN = 0
IV_MAX = 15


class IV(BaseModel):
    """個体値（Individual Values）.

    不変のValue Object。0-15の範囲で攻撃/防御/HPを表す。

    Attributes:
        attack: 攻撃個体値（0-15）
        defense: 防御個体値（0-15）
        stamina: HP個体値（0-15）
    """

    attack: int = Field(ge=IV_MIN, le=IV_MAX, description="攻撃個体値")
    defense: int = Field(ge=IV_MIN, le=IV_MAX, description="防御個体値")
    stamina: int = Field(ge=IV_MIN, le=IV_MAX, description="HP個体値")

    model_config = {"frozen": True}

    def __str__(self) -> str:
        """文字列表現を返す."""
        return f"{self.attack}/{self.defense}/{self.stamina}"

    @property
    def total(self) -> int:
        """合計値を返す."""
        return self.attack + self.defense + self.stamina

    @property
    def percentage(self) -> float:
        """パーセンテージを返す（0-100%）."""
        max_total = IV_MAX * 3
        return (self.total / max_total) * 100

    @property
    def is_perfect(self) -> bool:
        """100%個体かどうか."""
        return (
            self.attack == IV_MAX
            and self.defense == IV_MAX
            and self.stamina == IV_MAX
        )
